File: eniato/apps/accounts/domain/account.py
from typing import Optional

from lib.users.domain import User
from lib.helpers import local_currency
from .financial_institution import FinancialInstitution

from apps.accounts.constants import ACCOUNT_TYPE_DISPLAY_NAME


class Account:
    """"
    Account for a specific user

    Attrs:
        - 
    """

    def __init__(self, attributes={}):
        self.id: Optional[int] = attributes.get('id')
        self.user: User = attributes['user']
        self.financial_institution: FinancialInstitution = attributes['financial_institution']
        self.opening_balance: float = attributes['opening_balance']
        self.current_balance: float = attributes.get('current_balance')
        self.description: str = attributes['description']
        self.account_type: str = attributes['account_type']
        self.color: str = attributes['color']
        self.default: bool = attributes['default']
        self.active: bool = attributes['active']

    def set_id(self, id):
        self.id = id

    def set_default(self, default):
        self.default = default

    @property
    def display_account_type(self):
        return dict(ACCOUNT_TYPE_DISPLAY_NAME)[self.account_type]

    @property
    def display_opening_balance(self):
        return local_currency(self.opening_balance)

    @property
    def display_current_balance(self):
        return local_currency(self.current_balance)

File: eniato/vue-app/src/apps/account/domain/Account.js
export class Account {
  constructor (attributes) {
    this._id = attributes.id
    this._user = attributes.user
    this._financialInstitution = attributes.financialInstitution
    this._openingBalance = attributes.openingBalance
    this._currentBalance = attributes.currentBalance
    this._description = attributes.description
    this._accountType = attributes.accountType
    this._displayAccountType = attributes.displayAccountType
    this._color = attributes.color
    this._default = attributes.default
    this._active = attributes.active
  }

  get id () { return this._id }
  get user () { return this._user }
  get financialInstitution () { return this._financialInstitution }
  get openingBalance () { return this._openingBalance }
  get currentBalance () { return this._currentBalance }
  get description () { return this._description }
  get accountType () { return this._accountType }
  get displayAccountType () { return this._displayAccountType }
  get color () { return this._color }
  get default () { return this._default }
  get active () { return this._active }

  toDTO () {
    return {
      id: this._id,
      user: this._user,
      financialInstitution: this._financialInstitution,
      openingBalance: this._openingBalance,
      currentBalance: this._currentBalance,
      description: this._description,
      accountType: this._accountType,
      displayAccountType: this._displayAccountType,
      color: this._color,
      default: this._default,
      active: this._active
    }
  }
}
